bug-finding.ipynb

```diff
@@ -290,11 +434,86 @@
    "source": [
     "# Bug 3 analysis and response"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 139,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "r1: 38, r2: 311516, c1: 26932, c2: 284622, o11: 23, o12: 15, o21: 26909\n"
+     ]
+    }
+   ],
+   "source": [
+    "def count_ngram_collocations(x, w1, w2, l_size: int = 1, r_size: int = 1):\n",
+    "    lemmata = [t.lemma_ for t in x]\n",
+    "\n",
+    "    indexes = [i for i, lemma in enumerate(lemmata) if lemma == w1]\n",
+    "\n",
+    "    cooccurrences = 0\n",
+    "\n",
+    "    for i in indexes:\n",
+    "        left = max(i - l_size, 0)\n",
+    "        right = min(i + r_size + 1, len(lemmata))\n",
+    "        window = lemmata[left:right]\n",
+    "\n",
+    "        if w2 in window:\n",
+    "            cooccurrences += 1\n",
+    "\n",
+    "    return cooccurrences\n",
+    "\n",
+    "def count_ngram_non_collocations(x, w1, w2, l_size: int = 1, r_size: int = 1):\n",
+    "    lemmata = [t.lemma_ for t in x]\n",
+    "\n",
+    "    indexes = [i for i, lemma in enumerate(lemmata) if lemma == w1]\n",
+    "\n",
+    "    non_cooccurrences = 0\n",
+    "\n",
+    "    for i in indexes:\n",
+    "        left = max(i - l_size, 0)\n",
+    "        right = min(i + r_size + 1, len(lemmata))\n",
+    "        window = lemmata[left:right]\n",
+    "\n",
+    "        if w2 not in window:\n",
+    "            non_cooccurrences += 1\n",
+    "\n",
+    "    return non_cooccurrences\n",
+    "\n",
+    "r1 = len([t for t in df[\"nlp_docs\"].explode() if t.lemma_ == node])\n",
+    "r2 = len([t for t in df[\"nlp_docs\"].explode() if t.lemma_ != node])\n",
+    "c1 = len([t for t in df[\"nlp_docs\"].explode() if t.lemma_ == collocate])\n",
+    "c2 = len([t for t in df[\"nlp_docs\"].explode() if t.lemma_ != collocate])\n",
+    "\n",
+    "o11 = df[\"nlp_docs\"].apply(count_ngram_collocations, args=(node, collocate)).sum()\n",
+    "o12 = df[\"nlp_docs\"].apply(count_ngram_non_collocations, args=(node, collocate)).sum()\n",
+    "# o12 = r1 - o11\n",
+    "o21 = c1 - o11\n",
+    "\n",
+    "print(f\"r1: {r1}, r2: {r2}, c1: {c1}, c2: {c2}, o11: {o11}, o12: {o12}, o21: {o21}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For this last question, I have to admit that I was completely lost. After many countless attempts to fix the code by trial and error, I referenced the updated code in the week_04_inclass.ipynb, from which I copied and pasted the code above. (I hope this was allowed; I assumed this was \"open book\" like the other quizzes). My coding skills are still very, very basic, and I'm not able to produce or fully understand such complex codes yet. I thought it was better to use the class notebook and try to understand why it works rather than just give up and leave the question blank. \n",
```

Comment:
Yes, this is what I was hoping you would do! This is a totally valid way to learn to code.